File: 1.0/1.- Nanocrystalline Gold/HPC Code/src/dynamics.c
#include <math.h>
#include <complex.h>
#include <fftw3-mpi.h>
#include <gsl/gsl_randist.h>

#include "state.h"
#include "dynamics.h"

static double ipow (double x, int p)
{
    if (p == 1) return x;
    else return x * ipow (x, p-1);
}

static void normalize (state *s, fftw_complex *field)
{
    /* Normalize a fourier space field after fft */

    int ij;
    double norm_scale = 1.0 /(s->N * s->N);

    for (int i = 0; i < s->local_n1; i++)
    {
        for (int j = 0; j < s->N; j++)
        {
            ij = i*s->N + j;
            field[ij] *= norm_scale;
        }
    }

    MPI_Barrier (MPI_COMM_WORLD);

    return;
}

static double F_mix (double c, state *s)
{
    double epsilon = -4.0 + s->epsilon0*(s->sigma - s->sigma0);

    return s->omega*(
                c * log(2.0 * c) + (1 - c) * log(2.0 - 2.0 * c) +
                0.5 * epsilon * ipow(c - 0.5, 2)
            );
}

static double dF_mixdc (double c, state *s)
{
    return s->omega * log(c / (1.0 - c));
}

static void noise(state *s)
{
    fftw_complex c_scale, n_scale;
    c_scale = I*sqrt(s->kbT * s->Mc * s->dt / (s->dx * s->dx));
    n_scale = I*sqrt(s->kbT * s->Mn * s->dt / (s->dx * s->dx));

    int ij;

    for (int i = 0; i < s->local_n1; i++)
    {
        for (int j = 0; j < s->N; j++)
        {
            ij = i*s->N + j;
            s->fxin[ij] = sqrt(s->k2[ij]) * n_scale * gsl_ran_gaussian_ziggurat(s->rng, 1.0) +
                      I * sqrt(s->k2[ij]) * n_scale * gsl_ran_gaussian_ziggurat(s->rng, 1.0);

            s->fxic[ij] = sqrt(s->k2[ij]) * c_scale * gsl_ran_gaussian_ziggurat(s->rng, 1.0) +
                      I * sqrt(s->k2[ij]) * c_scale * gsl_ran_gaussian_ziggurat(s->rng, 1.0);
        }
    }

    MPI_Barrier (MPI_COMM_WORLD);

    return;
}

static void set_nonlinear (state *s)
{
    double third = 1.0/3.0;
    int ij;

    for (int i = 0; i < s->local_n0; i++)
    {
        for (int j = 0; j < s->N; j++)
        {
            ij = i*2*((s->N>>1) + 1) + j;

            s->nnl[ij]  = ipow(s->n[ij], 2)*( -0.5 * s->eta + third * s->n[ij]);
            s->nnl[ij] += F_mix (s->c[ij], s);
            s->nnl[ij] += -exp(-ipow(s->c[ij] - 1.0, 2)/(2.0*ipow(s->alphac, 2))) * s->Cn[ij];

            s->cnl[ij]  = (s->n[ij] + 1.0) *  dF_mixdc (s->c[ij], s);
            s->cnl[ij] += 0.5 * s->n[ij] * (s->c[ij] - 1.0) / ipow(s->alphac, 2) *
                          exp(-ipow(s->c[ij] - 1.0, 2)/(2.0*ipow(s->alphac, 2))) *
                          s->Cn[ij];
        }
    }

    MPI_Barrier (MPI_COMM_WORLD);

    return;
}

static void calccorr (state *s)
{
    int ij;
    double norm = 1.0/(s->N * s->N);

    for (int i = 0; i < s->local_n1; i++)
    {
        for (int j = 0; j < s->N; j++)
        {
            ij = i * s->N + j;
            s->fn[ij] *= norm;
            s->fCn[ij] = s->C[ij]*s->fn[ij];
        }
    }

    fftw_mpi_execute_dft_c2r (s->ifft_plan, s->fCn, s->Cn);

    return;
}

static void propagate (state *s)
{
    double epsilon = -4.0 + s->epsilon0 * (s->sigma - s->sigma0);
    double norm = 1.0 / (s->N * s->N);
    int ij;

    for (int i = 0; i < s->local_n1; i++)
    {
        for (int j = 0; j < s->N; j++)
        {
            ij = i * s->N + j;

            /* Do some normalization quickly */
            s->fc[ij] *= norm;
            s->fnnl[ij] *= norm;
            s->fcnl[ij] *= norm;

            s->fn[ij]  = 1.0 / (1.0 + s->dt * s->k2[ij]) *
                        (s->fn[ij] - s->dt * s->k2[ij] * s->fnnl[ij] + s->dt * s->fxin[ij]);

            s->fc[ij]  = 1.0 / (1.0 + s->dt * s->k2[ij] * (s->omega * epsilon + s->Wc * s->k2[ij])) *
                        (s->fc[ij] - s->dt * s->k2[ij] * s->fcnl[ij] + s->dt * s->fxic[ij]);
        }
    }

    MPI_Barrier (MPI_COMM_WORLD);

    return;
}

void step(state *s)
{
    /* Fourier transform fields */
    fftw_mpi_execute_dft_r2c (s->fft_plan, s->n, s->fn);
    fftw_mpi_execute_dft_r2c (s->fft_plan, s->c, s->fc);

    /* Calculate (C * n) and the nonlinear term */
    calccorr (s);
    set_nonlinear (s);

    /* Fourier transform the nonlinear terms */
    fftw_mpi_execute_dft_r2c (s->fft_plan, s->nnl, s->fnnl);
    fftw_mpi_execute_dft_r2c (s->fft_plan, s->cnl, s->fcnl);
    
    /* Make some noise and propagate the fields in fourier space */
    noise (s);
    propagate (s);

    /* Inverse fourier transform the propagated fields */
    fftw_mpi_execute_dft_c2r (s->ifft_plan, s->fn, s->n);
    fftw_mpi_execute_dft_c2r (s->ifft_plan, s->fc, s->c);

    /* Update the time and step #   */
    s->t += s->dt;
    s->step += 1;

    MPI_Barrier (MPI_COMM_WORLD);

    return;
}

File: Models/1.- Nanocrystalline Gold/Code/dynamics.jl
export step!

function ΔFmix(c::Float64, s::State)
	ϵ = -4.0 + s.ϵ₀*(s.σ - s.σ₀)
	return s.ω*(c*log(2.0c) + (1-c)*log(2.0*(1-c)) + 0.5*ϵ*(c-0.5)^2)
end

δΔFmixδc(c::Float64, s::State) = s.ω*(log(2.0c) - log(2.0*(1-c)))

function noise!(s::State)
	c_scale = im*√(s.Mc/(s.Δx^2*s.Δt))
	n_scale = im*√(s.Mₙ/(s.Δx^2*s.Δt))
	@threads for j in 1:s.N
		for i in 1:s.N>>1 + 1
			s.ξc[i, j] = index_to_k(s, i, j)*c_scale*complex(randn(), randn())
			s.ξₙ[i, j] = index_to_k(s, i, j)*n_scale*complex(randn(), randn())
		end
	end
	return nothing
end

function set_nonlinear!(s::State)
	third = 1/3
	for i in 1:s.N*s.N
		s.nₙₗ[i] = s.n[i]^2*(-0.5*s.η + third*s.χ*s.n[i]) + ΔFmix(s.c[i], s) - exp(-(s.c[i]-1.0)^2/(2s.αc^2))*s.C₂n[i]
		s.cₙₗ[i] = (s.n[i] + 1.)*δΔFmixδc(s.c[i], s) - 0.5*s.n[i]*(-(s.c[i]-1.0)/s.αc^2*exp(-(s.c[i]-1.0)^2/(2s.αc^2))*s.C₂n[i])
	end
end

function calccorr!(s::State)
	for i in 1:(s.N>>1+1)*s.N
        s.kC₂n[i] = s.C₂[i]*s.ñ[i]
    end
    A_ldiv_B!(s.C₂n, s.fftplan, s.kC₂n)
    return nothing
end

function step!(s::State)
	A_mul_B!(s.ñ, s.fftplan, s.n)
	A_mul_B!(s.c̃, s.fftplan, s.c)
	calccorr!(s)
	set_nonlinear!(s)
	A_mul_B!(s.ñₙₗ, s.fftplan, s.nₙₗ)
	A_mul_B!(s.c̃ₙₗ, s.fftplan, s.cₙₗ)
	noise!(s)
	ϵ = -4.0 + s.ϵ₀*(s.σ - s.σ₀)
	for i in 1:(s.N>>1 + 1)*s.N
		s.ñ[i] = 1.0/(1.0-s.Δt*s.∇²[i])*(s.ñ[i] + s.Δt*s.∇²[i]*s.ñₙₗ[i] + s.Δt*s.ξₙ[i])
		s.c̃[i] = 1.0/(1.0-s.Δt*s.∇²[i]*(s.ω*ϵ-s.Wc*s.∇²[i]))*(s.c̃[i] + s.Δt*s.∇²[i]*s.c̃ₙₗ[i] + s.Δt*s.ξc[i])
	end
	A_ldiv_B!(s.n, s.fftplan, s.ñ)
	A_ldiv_B!(s.c, s.fftplan, s.c̃)
	return nothing
end
